Run "pub global activate devtools" before overall_experience_test

Hixie opened the pull request:
This should reduce the chance of the test failing with timeouts while trying to launch devtools.

Intended to address https://github.com/flutter/flutter/issues/82382 and https://github.com/flutter/flutter/issues/81486.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

Hixie committed:
Run "pub global activate devtools" before overall_experience_test

This should reduce the chance of the test failing with timeouts while trying to launch devtools.

Hixie commented:
@CaseyHillers is there some way I can test this on devicelab a few times before landing? It's supposed to make a test less flaky but I've no idea how successful it will be.

CaseyHillers commented:
> @CaseyHillers is there some way I can test this on devicelab a few times before landing? It's supposed to make a test less flaky but I've no idea how successful it will be.

Do you want to run the entire devicelab test suite evaluate its impact, or do you just want to try it on a few prod tasks? I can schedule a few presubmit tasks with this change if you list which ones you're interested in.

Hixie commented:
It would just be a couple of specific tests. But since every presubmit in the world has failed on this PR let me deal with that first! ;-)

Hixie commented:
Wait, this isn't even a devtools test actually. What am I talking about. Never mind. Sorry @CaseyHillers.

Hixie commented:
cc @jonahwilliams I've run the tests a few times here (forced pushed the branch several times) and it's never flaked. Dunno how else to test this really.

jonahwilliams commented on packages/flutter_tools/lib/src/devtools_launcher.dart:
```diff
@@ -97,9 +97,9 @@ class DevtoolsServerLauncher extends DevtoolsLauncher {
-      }
-      if (!devToolsActive && !offline) {
-        devToolsActive = await _checkForActiveDevTools();
+        if (!devToolsActive) {
```
why this change? if we know that we can't reach pub, then re-activating devtools will almost always fail - unless it is already activated. but then `_checkForActiveDevTools` should have already come back as true

Hixie replied:
This is a no-op change. It's semantically identical, except in principle might save a couple of cycles on a redundant check if the compiler didn't notice that it was semantically identical.

zanderso commented on dev/bots/test.dart:
```diff
@@ -301,6 +301,20 @@ Future<void> _runIntegrationToolTests() async {
+  }
+  await runCommand(
+    pub,
+    <String>['global', 'activate', 'devtools'],
```
@jonahwilliams mentioned today that there can be a risk of inconsistent state if two instances of `pub` are running at the same time in the same pub cache. @jonahwilliams could this invocation race with the invocation in the flutter_tool? Also, due to a recent break, we'd like to pin devtools: https://github.com/flutter/flutter/pull/86124

```suggestion
    <String>['global', 'activate', 'devtools', '2.4.0'],
```

jonahwilliams replied:
As long as this command is run before we start any tools or other tests it should be OK. In the previous case pub global activate was running close to the same time as the flutter tool was started along side the host side driver test running update

Hixie replied:
I'll make this pin the pubtools version (but with a constant).

As Jonah says, this should be pretty safe. Nothing else should be running on the VM at this point.

zanderso replied:
I've got a PR going with a constant here https://github.com/flutter/flutter/pull/86130

Hixie replied:
Added a commit that pins with a `.version` file like other pins.

Hixie committed:
Pin devtools using a version file

Hixie commented:
I wasn't quite sure how to get `flutterRoot` without using globals, let me know if I missed something.

I used the real filesystem in the unit tests. If there's a better solution let me know.

jonahwilliams reviewed (approved):
LGTM

Hixie commented:
Yay, I actually got a flake here finally. Looks like it was caused by the `flutter` tool taking too long to build, so I also added running the `flutter` tool to the `test.dart` preparations.

Hixie committed:
Redirect stderr to stdout in test.dart

LUCI makes it much harder to manage output since it decouples stdout and stderr. To simplify our lives, this merges them together in test.dart output.

Hixie committed:
Reduce pub verbosity on LUCI

This reverts https://github.com/flutter/flutter/pull/80319 since the P0 issue is resolved now. The verbosity of pub triggered here is so high that most logs end up swamped with it.

Hixie committed:
Avoid precompiling the tool for the devicelab tests

Hixie commented:
@jonahwilliams You may wish to rereview this PR since I made a few changes since you LGTM'ed it.

jonahwilliams reviewed (approved):
Still LGTM
